From d13e6e44f5b57a33968c9ebb74e2b3bb12df7215 Mon Sep 17 00:00:00 2001
From: Mohit Sabharwal <mohit@cloudera.com>
Date: Fri, 29 Jul 2016 15:12:36 -0700
Subject: [PATCH 0720/1164] CDH-42737 : HIVE-13788 : hive msck listpartitions
 need to make use of directSQL instead of
 datanucleus

    Signed-off-by: Ashutosh Chauhan <hashutosh@apache.org>
    (cherry picked from commit 696104fab350a470a745530105d34f1015ded664)

    Change-Id: If45e9b93a5fed55a83c62998a03964afe96bbd83

Change-Id: Ibc7b58162e6412167a10ba392375147bf5c5ee75
---
 .../hive/ql/metadata/HiveMetaStoreChecker.java     |    6 +++++-
 .../hive/ql/optimizer/ppr/PartitionPruner.java     |    6 ++++--
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveMetaStoreChecker.java b/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveMetaStoreChecker.java
index fa0abad..33b9c58 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveMetaStoreChecker.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveMetaStoreChecker.java
@@ -37,6 +37,8 @@
 import org.apache.hadoop.hive.metastore.api.MetaException;
 import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
 import org.apache.hadoop.hive.ql.metadata.CheckResult.PartitionResult;
+import org.apache.hadoop.hive.ql.optimizer.ppr.PartitionPruner;
+import org.apache.hadoop.hive.ql.parse.PrunedPartitionList;
 import org.apache.thrift.TException;
 
 /**
@@ -196,8 +198,10 @@ void checkTable(String dbName, String tableName,
 
     if (table.isPartitioned()) {
       if (partitions == null || partitions.isEmpty()) {
+        PrunedPartitionList prunedPartList =
+        PartitionPruner.prune(table, null, conf, toString(), null);
         // no partitions specified, let's get all
-        parts = hive.getPartitions(table);
+        parts.addAll(prunedPartList.getPartitions());
       } else {
         // we're interested in specific partitions,
         // don't check for any others
diff --git a/ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartitionPruner.java b/ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartitionPruner.java
index cb99aec..d0bbfe8 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartitionPruner.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/optimizer/ppr/PartitionPruner.java
@@ -219,7 +219,7 @@ public static PrunedPartitionList prune(Table tab, ExprNodeDesc prunerExpr,
 
   private static PrunedPartitionList getAllPartsFromCacheOrServer(Table tab, String key, boolean unknownPartitions,
     Map<String, PrunedPartitionList> partsCache)  throws SemanticException {
-    PrunedPartitionList ppList = partsCache.get(key);
+    PrunedPartitionList ppList = partsCache == null ? null : partsCache.get(key);
     if (ppList != null) {
       return ppList;
     }
@@ -230,7 +230,9 @@ private static PrunedPartitionList getAllPartsFromCacheOrServer(Table tab, Strin
       throw new SemanticException(e);
     }
     ppList = new PrunedPartitionList(tab, parts, null, unknownPartitions);
-    partsCache.put(key, ppList);
+    if (partsCache != null) {
+      partsCache.put(key, ppList);
+    }
     return ppList;
   }
 
-- 
1.7.9.5

